### PR TITLE
CRB-301 exit if min_peers is not sustained

### PR DESCRIPTION
**Pull request by CAGS295**
Go to the ticket for a description.

### CODE REVIEW

File: validator/sawtooth_validator/server/core.py [hunk 1]
```diff
@@ -452,6 +452,9 @@ def start(self):
         else:
             self._start()
 
+    class RebootException(Exception):
+        ...
```
Comment (atodorov) [inline validator/sawtooth_validator/server/core.py]: `...` isn't valid Python. This will lead to SyntaxError. Use `pass`.

**Review by atodorov — CHANGES_REQUESTED** [review]

**Review by atodorov — CHANGES_REQUESTED** [review]
Doesn't make things better. See ticket for more info and logs.

**Comment by nathanwhit** [comment]
This will trigger a graceful exit, but in my experience a graceful exit is not guaranteed to actually terminate, depending on the state of zeroMQ sockets. The disconnect messages are sent immediately when a graceful exit is triggered, so forcefully exiting after a bit should be okay. 

You could maybe try a timer or something in the exit logic that forces an exit if the graceful exit doesn't complete in X seconds.

**Comment by nathanwhit** [comment]
@atodorov Pushed a new commit, give it a try and let me know how it goes

**Review by atodorov — APPROVED** [review]
@nathanwhit , @CAGS295 

```
[2022-01-12 19:52:58.037 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.037 ERROR    cli] 
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/sawtooth_validator/server/cli.py", line 299, in main
    validator.start()
  File "/usr/lib/python3/dist-packages/sawtooth_validator/server/core.py", line 454, in start
    self._start()
  File "/usr/lib/python3/dist-packages/sawtooth_validator/server/core.py", line 481, in _start
    raise self.RebootException()
sawtooth_validator.server.core.Validator.RebootException
[2022-01-12 19:52:58.039 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.041 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.042 ERROR    proxy] Settings from block b62f9590e71049c971ebadd42cd9eb4ca477fa8f2b2360c30d5268720a04f1f203ac773ebdcc8cd4b2ae6b83b638ee060b3318a8b1b36f6d46ebc4b0ed7b944c requested, but root hash 2284aaab059f6094f2ac913511eceb7d084ed3bc2e9fdcaa48248fb7cb5cc733 was missing. Returning no setting values.
[2022-01-12 19:52:58.043 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.044 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.045 DEBUG    dispatch] Removed send_message function for connection ServerThread
[2022-01-12 19:52:58.046 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.047 DEBUG    dispatch] Removed send_last_message function for connection ServerThread
[2022-01-12 19:52:58.050 DEBUG    ffi] [src/journal/block_scheduler.rs: 119] During block scheduling, previous block b62f9590e71049c971ebadd42cd9eb4ca477fa8f2b2360c30d5268720a04f1f203ac773ebdcc8cd4b2ae6b83b638ee060b3318a8b1b36f6d46ebc4b0ed7b944c is pending, adding block 0c1484c2891e7b5ed68aa54a3bad9db23309d6f8dcf8b3d51a752017463151036693e1c9915d2d084ee5a8df4dcd1b45102bbdbd56347cca27c401ebd42f201c to pending
[2022-01-12 19:52:58.051 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.051 DEBUG    dispatch] Removed send_message function for connection OutboundConnectionThread-tcp://20.188.27.55:8800
[2022-01-12 19:52:58.053 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.053 DEBUG    dispatch] Removed send_last_message function for connection OutboundConnectionThread-tcp://20.188.27.55:8800
[2022-01-12 19:52:58.054 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.054 DEBUG    dispatch] Removed send_message function for connection OutboundConnectionThread-tcp://40.112.252.222:8800
[2022-01-12 19:52:58.055 ERROR    proxy] Settings from block 0c1484c2891e7b5ed68aa54a3bad9db23309d6f8dcf8b3d51a752017463151036693e1c9915d2d084ee5a8df4dcd1b45102bbdbd56347cca27c401ebd42f201c requested, but root hash 2284aaab059f6094f2ac913511eceb7d084ed3bc2e9fdcaa48248fb7cb5cc733 was missing. Returning no setting values.
[2022-01-12 19:52:58.057 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.057 DEBUG    dispatch] Removed send_last_message function for connection OutboundConnectionThread-tcp://40.112.252.222:8800
[2022-01-12 19:52:58.058 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.058 DEBUG    dispatch] Removed send_message function for connection OutboundConnectionThread-tcp://54.185.193.31:8800
[2022-01-12 19:52:58.059 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.060 DEBUG    dispatch] Removed send_last_message function for connection OutboundConnectionThread-tcp://54.185.193.31:8800
[2022-01-12 19:52:58.061 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.067 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.067 DEBUG    ffi] [src/journal/block_scheduler.rs: 119] During block scheduling, previous block 0c1484c2891e7b5ed68aa54a3bad9db23309d6f8dcf8b3d51a752017463151036693e1c9915d2d084ee5a8df4dcd1b45102bbdbd56347cca27c401ebd42f201c is pending, adding block 4f88a6ebcc10b88365bf01d0ff17e63e08577d284278e7e57e5a25a43e5ecdd93b0436d2240196b674041c0cd4c0819d24a38dc28968eaaa77d09c4ccb383f23 to pending
[2022-01-12 19:52:58.068 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.069 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.070 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.071 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.071 DEBUG    dispatch] Removed send_message function for connection ServerThread
[2022-01-12 19:52:58.072 ERROR    proxy] Settings from block 4f88a6ebcc10b88365bf01d0ff17e63e08577d284278e7e57e5a25a43e5ecdd93b0436d2240196b674041c0cd4c0819d24a38dc28968eaaa77d09c4ccb383f23 requested, but root hash 2284aaab059f6094f2ac913511eceb7d084ed3bc2e9fdcaa48248fb7cb5cc733 was missing. Returning no setting values.
[2022-01-12 19:52:58.072 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.072 DEBUG    dispatch] Removed send_last_message function for connection ServerThread
[2022-01-12 19:52:58.073 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.073 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.074 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.076 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.077 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.078 DEBUG    dispatch] Removed send_message function for connection ServerThread
[2022-01-12 19:52:58.078 DEBUG    ffi] [src/journal/block_scheduler.rs: 119] During block scheduling, previous block 4f88a6ebcc10b88365bf01d0ff17e63e08577d284278e7e57e5a25a43e5ecdd93b0436d2240196b674041c0cd4c0819d24a38dc28968eaaa77d09c4ccb383f23 is pending, adding block 67f03855622d3797a52c4e5a7c68820be8db0c6fabdb1e9a8066519191bd20165684c219abdc39e16e8e19f91554a8e3330c3a49b0fa6f1644e9297ad296d17a to pending
[2022-01-12 19:52:58.078 DEBUG    dispatch] Removed send_last_message function for connection ServerThread
[2022-01-12 19:52:58.078 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.080 WARNING  dispatch] Can't send message CONSENSUS_BLOCKS_GET_RESPONSE back to 65751933b3b8951f1ec60fb4b2cb45a0980d662c6beb1ad731d84021e0693e1f0ca470c5cd154d9bc35ed355c3d23b1ac79e498bf880d2514995d0f7f31816ac because connection ServerThread not in dispatcher
[2022-01-12 19:52:58.082 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.084 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.084 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.085 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.085 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.086 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.087 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.087 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.088 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.089 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.090 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.091 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.091 WARNING  (unknown file)] [src/journal/block_validator.rs: 293] Error during block validation: BlockValidationError("During validate_on_chain_rules, error creating settings view: NotFound(\"ff686b84cf07d9923bd5b7f1d6f2426a26b969a5d756ba189615dbc287211cb3\")")
[2022-01-12 19:52:58.144 WARNING  (unknown file)] [src/journal/publisher.rs: 598] PublisherThread exiting
[2022-01-12 19:52:58.682 DEBUG    gossip] Peers are: ['tcp://54.185.193.31:8800', 'tcp://20.113.27.206:8800']. Unpeered candidates count: 4.
[2022-01-12 19:52:58.682 DEBUG    gossip] endpoint tcp://20.188.27.55:8800 has 779351552eac9c86e938c12a55e708b665a7b16d85abe95acbbcb74d148a195cd00595a8b1cb554d488ecbf8f8d63a06fe0c2fb3ec70994858f4d042d3bf3567 an outstanding connection
[2022-01-12 19:53:08.102 INFO     core] Timed out waiting for threads to stop, exiting forcefully
[2022-01-12 19:53:08.102 ERROR    (unknown file)] error executing main
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/sawtooth_validator/server/cli.py", line 299, in main
    validator.start()
  File "/usr/lib/python3/dist-packages/sawtooth_validator/server/core.py", line 454, in start
    self._start()
  File "/usr/lib/python3/dist-packages/sawtooth_validator/server/core.py", line 481, in _start
    raise self.RebootException()
sawtooth_validator.server.core.Validator.RebootException

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/sawtooth_validator/server/cli.py", line 311, in main
    sys.exit(1)
SystemExit: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/sawtooth_validator/server/cli.py", line 315, in main
    validator.stop()
  File "/usr/lib/python3/dist-packages/sawtooth_validator/server/core.py", line 536, in stop
    sys.exit()
SystemExit
ubuntu@cc2-azure-africa:~/CreditcoinDockerCompose-2.0$ docker ps -a
CONTAINER ID   IMAGE                                            COMMAND                  CREATED          STATUS          PORTS                                                                                            NAMES
89f8c1b3c975   gluwa/creditcoin-processor-priv:testing-2.0-25   "./bin/ccprocessor-r…"   13 minutes ago   Up 13 minutes                                                                                                    creditcoin-processor
147eadf5cde9   hyperledger/sawtooth-settings-tp:1.2             "settings-tp -C tcp:…"   13 minutes ago   Up 13 minutes   4004/tcp                                                                                         sawtooth-settings-tp
fd7533e9d187   gluwa/creditcoin-consensus-priv:testing-2.0-26   "ccconsensus -E tcp:…"   13 minutes ago   Up 52 seconds                                                                                                    consensus
6e53d81e1cc7   gluwa/sawtooth-rest-api-priv:testing-2.0-27      "sawtooth-rest-api -…"   13 minutes ago   Up 13 minutes   4004/tcp, 0.0.0.0:8008->8008/tcp, :::8008->8008/tcp                                              sawtooth-rest-api
c0d56a45c2e7   gluwa/creditcoin-gateway-priv:testing-2.0-22     "./ccgateway -v"         13 minutes ago   Up 13 minutes   0.0.0.0:55555->55555/tcp, :::55555->55555/tcp                                                    creditcoin-gateway
c182c3fb1563   gluwa/creditcoin-validator-priv:testing-2.0-27   "bash -c 'if [[ ! -f…"   13 minutes ago   Up 5 minutes    0.0.0.0:4004->4004/tcp, :::4004->4004/tcp, 0.0.0.0:8800->8800/tcp, :::8800->8800/tcp, 5050/tcp   sawtooth-validator
```


You can see that we timed-out waiting for threads to stop and exited. Validator container was restarted and subsequently consensus container was also restarted, see different uptimes. ATM I can see both containers up and running so I am approving this PR.

**Comment by nathanwhit** [comment]
@atodorov Added another commit. Before we were blocking to wait on thread pools before exiting, now if we're rebooting we won't block.

**Review by seandeneen-gluwa — APPROVED** [review]

**Pull request merged**